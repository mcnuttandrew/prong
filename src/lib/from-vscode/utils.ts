import * as Json from "jsonc-parser";
import { JSONSchemaRef, JSONSchema } from "../JSONSchemaTypes";
import { ASTNode } from "./parser";
import * as nls from "vscode-nls";
let localize = nls.loadMessageBundle();

export function isNumber(val: any): val is number {
  return typeof val === "number";
}

export function isDefined(val: any): val is object {
  return typeof val !== "undefined";
}

export function isBoolean(val: any): val is boolean {
  return typeof val === "boolean";
}

export function isString(val: any): val is string {
  return typeof val === "string";
}

export function getNodeValue(node: ASTNode): any {
  return Json.getNodeValue(node);
}

export function asSchema(
  schema: JSONSchemaRef | undefined
): JSONSchema | undefined {
  if (isBoolean(schema)) {
    return schema ? {} : { not: {} };
  }
  return schema;
}

export function contains(
  node: ASTNode,
  offset: number,
  includeRightBound = false
): boolean {
  return (
    (offset >= node.offset && offset < node.offset + node.length) ||
    (includeRightBound && offset === node.offset + node.length)
  );
}

export function equals(one: any, other: any): boolean {
  if (one === other) {
    return true;
  }
  if (
    one === null ||
    one === undefined ||
    other === null ||
    other === undefined
  ) {
    return false;
  }
  if (typeof one !== typeof other) {
    return false;
  }
  if (typeof one !== "object") {
    return false;
  }
  if (Array.isArray(one) !== Array.isArray(other)) {
    return false;
  }

  var i: number, key: string;

  if (Array.isArray(one)) {
    if (one.length !== other.length) {
      return false;
    }
    for (i = 0; i < one.length; i++) {
      if (!equals(one[i], other[i])) {
        return false;
      }
    }
  } else {
    var oneKeys: string[] = [];

    for (key in one) {
      oneKeys.push(key);
    }
    oneKeys.sort();
    var otherKeys: string[] = [];
    for (key in other) {
      otherKeys.push(key);
    }
    otherKeys.sort();
    if (!equals(oneKeys, otherKeys)) {
      return false;
    }
    for (i = 0; i < oneKeys.length; i++) {
      if (!equals(one[oneKeys[i]], other[oneKeys[i]])) {
        return false;
      }
    }
  }
  return true;
}

export enum ErrorCode {
  Undefined = 0,
  EnumValueMismatch = 1,
  Deprecated = 2,
  UnexpectedEndOfComment = 0x101,
  UnexpectedEndOfString = 0x102,
  UnexpectedEndOfNumber = 0x103,
  InvalidUnicode = 0x104,
  InvalidEscapeCharacter = 0x105,
  InvalidCharacter = 0x106,
  PropertyExpected = 0x201,
  CommaExpected = 0x202,
  ColonExpected = 0x203,
  ValueExpected = 0x204,
  CommaOrCloseBacketExpected = 0x205,
  CommaOrCloseBraceExpected = 0x206,
  TrailingComma = 0x207,
  DuplicateKey = 0x208,
  CommentNotPermitted = 0x209,
  SchemaResolveError = 0x300,
}
