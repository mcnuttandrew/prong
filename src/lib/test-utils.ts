import { SyntaxNode } from "@lezer/common";
import { parser } from "@lezer/json";
// could also write one that is find by text content?
export function findNodeByLocation(
  text: string,
  from: number,
  to: number
): SyntaxNode | null {
  let foundNode: SyntaxNode | null = null;
  parser.parse(text).iterate({
    enter: (node) => {
      // console.log(node.from, node.to, node.node.type);
      if (node.from === from && node.to === to) {
        foundNode = node.node as unknown as SyntaxNode;
      }
    },
  });
  return foundNode;
}

export function findNodeByText(
  text: string,
  matchText: string
): SyntaxNode | null {
  let foundNode: SyntaxNode | null = null;
  parser.parse(text).iterate({
    enter: (node) => {
      // console.log(node.from, node.to, node.node.type);
      const testText = text.slice(node.from, node.to);
      if (testText === matchText) {
        foundNode = node.node as unknown as SyntaxNode;
      }
    },
  });
  return foundNode;
}
