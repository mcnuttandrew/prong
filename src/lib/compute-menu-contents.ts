import isequal from "lodash.isequal";
import { SyntaxNode } from "@lezer/common";
import { EditorView } from "@codemirror/view";
import * as Json from "jsonc-parser";

import { codeString, MenuEvent } from "./utils";
import { SchemaMap } from "../components/Editor";

type JSONSchema = any;

export type MenuRow = { label: string; element: MenuElement };
export type MenuElement =
  | { type: "display"; content: string }
  | { type: "button"; content: string; onSelect: MenuEvent }
  | {
      type: "row";
      label?: string;
      element: MenuElement[];
      direction: "horizontal" | "vertical";
    }
  | { label: "CUSTOM"; element: JSX.Element; type: "projection" };

//   TODO revisit to see if all are necessary
interface ComponentProps {
  content: JSONSchema;
  parsedContent: any;
  //   parentType: string; // todo this type can be improved
}
type Component = (props: ComponentProps) => MenuElement;
type componentContainer = Record<string, Component>;

const EnumPicker: Component = (props) => {
  const { content } = props;
  return {
    type: "row",
    direction: "vertical",
    element: (content.enum as any[]).map((val: string) => ({
      type: "button",
      content: val,
      onSelect: { type: "simpleSwap", payload: `"${val}"` },
    })),
  };
};

function simpleFillout(content: JSONSchema) {
  const simpleTypes: Record<string, any> = {
    string: "",
    object: {},
    number: 0,
    boolean: true,
  };
  if (content.type in simpleTypes) {
    return simpleTypes[content.type];
  } else if (content.anyOf && content.anyOf.length) {
    const childTypes = content.anyOf.map((x: any) => x.type);
    return childTypes.every((x: string) => childTypes[0] === x) &&
      childTypes[0] in simpleTypes
      ? simpleTypes[childTypes[0]]
      : null;
  } else {
    return null;
  }
}

const ObjPicker: Component = (props) => {
  const { content, parsedContent } = props;
  const currentKeys = new Set(Object.keys(parsedContent || {}));
  return {
    type: "row",
    direction: "horizontal",
    element: [
      { type: "display", content: "add fields" },
      ...(Object.entries(content.properties || {})
        .filter((x) => !currentKeys.has(x[0]))
        .map(([content, prop]) => ({
          type: "button",
          content,
          onSelect: {
            type: "addObjectKey",
            payload: { key: content, value: simpleFillout(prop) as any },
          },
        })) as MenuElement[]),
    ],
  };
};

// TODO flatten nested anyOfs and remove duplicates
function flattenAnyOf(content: JSONSchema) {
  if (!content || !content.anyOf) {
    return content;
  }
  return content.anyOf.reduce(
    (acc: any[], row: any) => acc.concat(row.anyOf ? flattenAnyOf(row) : row),
    []
  );
}

function removeDupsInAnyOf(content: JSONSchema[]) {
  return content.filter((row, idx) =>
    content.slice(idx + 1).every((innerRow) => !isequal(row, innerRow))
  );
}

function bundleConstsToEnum(content: JSONSchema[]) {
  const consts = content.filter((x) => x.const);
  const nonConsts = content.filter((x) => !x.const);
  return [...nonConsts, { enum: consts.map((x) => x.const) }];
}

function generateValueForObjProp(prop: any) {
  const simple: any = { number: 0, string: "", object: {} };
  if (prop.enum) {
    return prop.enum[0];
  } else if (simple.hasOwnProperty(prop.type)) {
    return simple[prop.type];
  } else {
    console.log("not covered", prop.type, simple[prop.type]);
    return null;
  }
}
// what do these exisit
const addToSet = (set: Set<string>, key: string) =>
  new Set([...Array.from(set), key]);

const removeFromSet = (set: Set<string>, key: string) =>
  new Set(Array.from(set).filter((x) => x !== key));

function AnyOfObjOptionalFieldPicker(
  content: JSONSchema,
  containerIdx: number
): MenuElement[] {
  const requiredProps = new Set<string>(
    Array.isArray(content.required) ? content.required : []
  );

  //   const [selectedOptions, setSelectedOptions] = useState<Set<string>>(
  //     new Set(Array.from(requiredProps))
  //   );
  return [
    content.$$labeledType && {
      type: "display",
      content: content.$$labeledType,
    },
    // {type: 'button', content: }
  ].filter((x) => x);
  //   return (
  //     <div style={{ display: "flex", flexDirection: "column" }}>
  //     //   {content.$$labeledType && (
  //     //     <span style={{ fontSize: "9px" }}>{content.$$labeledType}</span>
  //     //   )}
  //       <div className="flex">
  //         <button
  //           onClick={() => {
  //             const val = Object.fromEntries(
  //               Array.from(selectedOptions).map((key) => {
  //                 return [key, generateValueForObjProp(content.properties[key])];
  //               })
  //             );
  //             eventDispatch({ type: "simpleSwap", payload: JSON.stringify(val) });
  //           }}
  //         >
  //           Switch to obj with
  //         </button>
  //         // <div>
  //         //   {Object.keys(content?.properties || {}).map((key) => {
  //         //     const selected = selectedOptions.has(key);
  //         //     const required = requiredProps.has(key);
  //         //     const name = `${key}-${containerIdx}`;
  //         //     const onChange = required
  //         //       ? () => {}
  //         //       : () =>
  //         //           selected
  //         //             ? setSelectedOptions(removeFromSet(selectedOptions, key))
  //         //             : setSelectedOptions(addToSet(selectedOptions, key));
  //         //     return (
  //         //       <span>
  //         //         <label htmlFor={name}>{key}</label>
  //         //         <input
  //         //           type="checkbox"
  //         //           checked={required || selected}
  //         //           onChange={onChange}
  //         //           name={name}
  //         //         />
  //         //       </span>
  //         //     );
  //         //   })}
  //         // </div>
  //       </div>
  //     </div>
  //   );
}

function AnyOfArray(content: JSONSchema, idx: number): MenuElement[] {
  //   const [numElements, setNumbElements] = useState<number>(1);
  const numElements = 5;
  const arrayTypeDefaults: any = {
    boolean: true,
    string: "",
    number: 0,
    object: {},
    array: [],
  };
  const arrayType = content?.items?.type;
  const sliderName = `numElements${idx}`;
  return [
    !arrayType && {
      type: "button",
      content: "Switch to array",
      onSelect: { type: "simpleSwap", payload: "[]" },
    },
    arrayType && {
      type: "button",
      content: `Switch to array of ${numElements} ${JSON.stringify(
        arrayTypeDefaults[arrayType]
      )}s`,
      onSelect: {
        type: "simpleSwap",
        payload: JSON.stringify(
          [...new Array(numElements)].map(() => arrayTypeDefaults[arrayType])
        ),
      },
    },
    { type: "display", content: "TODO SLIDER" },
  ].filter((x) => x) as MenuElement[];
  //   return (
  //     <div className="flex">
  //   {!arrayType && (
  //     <button
  //       onClick={() => eventDispatch({ type: "simpleSwap", payload: "[]" })}
  //     >
  //       Switch to array
  //     </button>
  //   )}
  //       {arrayType && (
  //         <div>
  //           <button
  //             onClick={() =>
  //               eventDispatch({
  //                 type: "simpleSwap",
  //                 payload: JSON.stringify(
  //                   [...new Array(numElements)].map(
  //                     () => arrayTypeDefaults[arrayType]
  //                   )
  //                 ),
  //               })
  //             }
  //           >
  //             {`Switch to array of ${numElements} ${JSON.stringify(
  //               arrayTypeDefaults[arrayType]
  //             )}s`}
  //           </button>
  //           <div className="flex">
  //             <input
  //               type="range"
  //               id={sliderName}
  //               name={sliderName}
  //               min="0"
  //               max="11"
  //               value={numElements}
  //               onChange={(e) => setNumbElements(Number(e.target.value))}
  //             />
  //             <label htmlFor={sliderName}>num</label>
  //           </div>
  //         </div>
  //       )}
  //     </div>
  //   );
}

const AnyOfPicker: Component = (props) => {
  const { content } = props;
  const simpleType = new Set(["string", "number", "boolean", "null"]);
  const simpleTypeMap: Record<string, string> = {
    string: '""',
    number: "0",
    boolean: "true",
    null: "null",
  };
  const anyOptions = bundleConstsToEnum(
    removeDupsInAnyOf(flattenAnyOf(content))
  );
  return {
    type: "row",
    direction: "horizontal",
    element: anyOptions.flatMap((opt, idx) => {
      return [
        opt.description && { type: "display", content: opt.description },
        ...(opt.enum || []).map((val: string) => {
          return {
            type: "button",
            content: val,
            onSelect: { type: "simpleSwap", payload: `"${val}"` },
          };
        }),
        ...(opt.type === "object" ? AnyOfObjOptionalFieldPicker(opt, idx) : []),
        ...(opt.type === "array" ? AnyOfArray(opt, idx) : []),
        simpleType.has(opt.type) && {
          direction: "vertical",
          element: [
            opt.$$labeledType && {
              type: "display",
              content: opt.$$labeledType,
            },
            {
              type: "button",
              content: `switch to ${opt.type}`,
              onSelect: {
                type: "simpleSwap",
                // payload: "null"
                payload: simpleTypeMap[opt.type],
              },
            },
          ].filter((x) => x) as MenuElement[],
        },
      ];
    }),
  };
  // );
};

// function contentDescriber(description: string | null) {
//   if (!description) {
//     return null;
//   }
//   return (
//     <div style={{ maxHeight: "200px", overflowY: "auto" }}>
//       <ReactMarkdown>{description}</ReactMarkdown>
//     </div>
//   );
// }

const GenericComponent: Component = (props) => {
  return { type: "display", content: "hi generic" };
};

const PropertyNameComponent: Component = (props) => {
  const { parsedContent } = props;
  return {
    type: "row",
    direction: "horizontal",
    element: [
      { type: "display", content: `${parsedContent}` },
      {
        type: "button",
        content: "remove key",
        onSelect: { type: "removeObjectKey" },
      },
    ],
  };
};

const ObjectComponent: Component = (props) => {
  //   const [keyVal, setKeyVal] = React.useState("");
  //   const [valueVal, setValueVal] = React.useState("");
  //   const newKeyVal = (key: string, value: any) => ({
  //     type: "addObjectKey",
  //     payload: { key, value },
  //   });
  return { type: "display", content: "add field" };
  //   return (
  //     <div>
  //       <div>Add field</div>
  //       <div>
  //         <span>Key</span>
  //         <input value={keyVal} onChange={(e) => setKeyVal(e.target.value)} />
  //         <span>Value</span>
  //         <input value={valueVal} onChange={(e) => setValueVal(e.target.value)} />
  //         <button onClick={() => newKeyVal(keyVal, valueVal)}>
  //           Add new entry
  //         </button>
  //       </div>
  //     </div>
  //   );
};

const ParentIsPropretyComponent: Component = (props) => {
  return { type: "display", content: "hi property" };
};
const nullInsert: MenuEvent = { type: "nullEvent" };
const ParentIsArrayComponent: Component = (props) => {
  return {
    direction: "horizontal",
    type: "row",
    element: [
      {
        type: "button",
        content: "Remove Item",
        onSelect: { type: "removeElementFromArray" },
      },
      {
        direction: "vertical",
        type: "row",
        element: [
          {
            type: "button",
            content: "Move item forward",
            onSelect: { ...nullInsert },
          },
          {
            type: "button",
            content: "Move item backward",
            onSelect: { ...nullInsert },
          },
          {
            type: "button",
            content: "Set item as Last",
            onSelect: { ...nullInsert },
          },
          {
            type: "button",
            content: "Set item as First",
            onSelect: { ...nullInsert },
          },
        ],
      },
    ],
  };
};

const ArrayComponent: Component = () => ({
  type: "display",
  content: "hi array",
});

const StringComponent: Component = () => ({
  type: "display",
  content: "hi string",
});

const NumberComponent: Component = () => ({
  type: "display",
  content: "hi number",
});

const BooleanComponent: Component = () => ({
  type: "display",
  content: "hi boolean",
});

const NullComponent: Component = () => ({
  type: "display",
  content: "hi null",
});

const menuSwitch: componentContainer = {
  EnumPicker,
  ObjPicker,
  AnyOfPicker,
  GenericComponent,
};
const typeBasedComponents: componentContainer = {
  Object: ObjectComponent,
  PropertyName: PropertyNameComponent,
  Array: ArrayComponent,
  String: StringComponent,
  Number: NumberComponent,
  BooleanComponent: BooleanComponent,
  Null: NullComponent,
};
const parentResponses: componentContainer = {
  Property: ParentIsPropretyComponent,
  Array: ParentIsArrayComponent,
  // TODO: do i need to fill in all the other options for this?
};

function simpleParse(content: any) {
  try {
    return Json.parse(content);
  } catch (e) {
    return {};
  }
}
function retargetToAppropriateNode(node: SyntaxNode, schemaMap: SchemaMap) {
  let targetNode = node;
  if (node.type.name === "{" || node.type.name === "}") {
    targetNode = node.parent!;
  }
  // else if (node.type.name === "PropertyName") {
  //   targetNode = node.nextSibling!;
  // }

  const from = targetNode.from;
  const to = targetNode.to;

  let schemaChunk: JSONSchema = schemaMap[`${from}-${to}`];
  if (schemaChunk?.length > 1) {
    return { anyOf: schemaChunk };
  } else if (schemaChunk?.length === 1) {
    return schemaChunk[0];
  }

  return schemaChunk;
}

export function generateMenuContent(
  //   view: EditorView,
  currentCodeSlice: string,
  syntaxNode: SyntaxNode,
  schemaMap: SchemaMap
): MenuRow[] {
  //   const currentCodeSlice = codeString(view, syntaxNode.from, syntaxNode.to);
  const schemaChunk = retargetToAppropriateNode(syntaxNode, schemaMap);

  const type = syntaxNode.type.name;
  let typeBasedProperty: Component | null = null;
  if (typeBasedComponents[type]) {
    typeBasedProperty = typeBasedComponents[type];
  } else if (!schemaChunk) {
    console.log("missing imp for", type);
  } else if (typeBasedProperty && !typeBasedProperty[type]) {
    console.log("missing type imp for", type);
  }

  const parentType = syntaxNode.parent!.type.name;
  let contentBasedItem: Component | null = null;
  // TODO work through options listed in the validate wip
  if (schemaChunk && schemaChunk.enum) {
    contentBasedItem = menuSwitch.EnumPicker;
  } else if (schemaChunk && schemaChunk.type === "object") {
    contentBasedItem = menuSwitch.ObjPicker;
  } else if (schemaChunk && schemaChunk.anyOf) {
    contentBasedItem = menuSwitch.AnyOfPicker;
  }

  const parsedContent = simpleParse(currentCodeSlice);

  const content: MenuRow[] = [
    schemaChunk?.description && {
      label: "DESC",
      element: {
        type: "display",
        content: schemaChunk.description,
      },
    },
    schemaChunk &&
      !!contentBasedItem && {
        label: "CONTENT",
        element: contentBasedItem({
          parsedContent,
          //   parentType,
          //   ...props,
          content: schemaChunk,
        }),
      },
    typeBasedProperty && {
      label: "TYPE",
      element: typeBasedProperty({
        parsedContent,
        // parentType,
        // ...props,
        content: schemaChunk,
      }),
    },
    parentResponses[parentType] && {
      label: "PARENT",
      element: parentResponses[parentType]({
        parsedContent,
        // parentType,
        // ...props,
        content: schemaChunk,
      }),
    },
  ].filter((x) => x);
  return content;
}
