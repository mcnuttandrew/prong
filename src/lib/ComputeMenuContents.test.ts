import { generateMenuContent } from "./compute-menu-contents";
import { findNodeByText } from "./test-utils";
import { createNodeMap } from "./utils";

const schema = {
  $id: "https://example.com/arrays.schema.json",
  $schema: "https://json-schema.org/draft/2020-12/schema",
  description: "A representation of a person, company, organization, or place",
  type: "object",
  properties: {
    fruits: {
      type: "array",
      items: {
        type: "string",
      },
    },
    vegetables: {
      type: "array",
      items: { $ref: "#/$defs/veggie" },
    },
  },
  $defs: {
    veggie: {
      type: "object",
      required: ["veggieName", "veggieLike"],
      properties: {
        veggieName: {
          type: "string",
          description: "The name of the vegetable.",
        },
        veggieLike: {
          type: "boolean",
          description: "Do I like this vegetable?",
        },
      },
    },
  },
};

const exampleData = `{
  "fruits": [ "apple", "orange", "pear" ],
  "vegetables": [
    {
      "veggieName": "potato",
      "veggieLike": true
    },
    {
      "veggieName": "broccoli",
      "veggieLike": false
    }
  ]
}`;

test("generateMenuContent", async () => {
  const veg = findNodeByText(exampleData, `"vegetables"`)!;
  const fAndVNodeMap = await createNodeMap(schema, exampleData);
  expect(fAndVNodeMap).toMatchSnapshot();
  const menuContent = generateMenuContent(veg, fAndVNodeMap, exampleData);
  expect(menuContent).toMatchSnapshot();
});
