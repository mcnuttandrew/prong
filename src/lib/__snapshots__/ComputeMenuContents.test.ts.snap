// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateMenuContent 1`] = `
Object {
  "0-206": Array [
    Object {
      "$defs": Object {
        "veggie": Object {
          "properties": Object {
            "veggieLike": Object {
              "$$labeledType": "veggieLike",
              "description": "Do I like this vegetable?",
              "type": "boolean",
            },
            "veggieName": Object {
              "$$labeledType": "veggieName",
              "description": "The name of the vegetable.",
              "type": "string",
            },
          },
          "required": Array [
            "veggieName",
            "veggieLike",
          ],
          "type": "object",
        },
      },
      "$id": "https://example.com/arrays.schema.json",
      "$schema": "https://json-schema.org/draft/2020-12/schema",
      "description": "A representation of a person, company, organization, or place",
      "properties": Object {
        "fruits": Object {
          "$$labeledType": "fruits",
          "items": Object {
            "type": "string",
          },
          "type": "array",
        },
        "vegetables": Object {
          "$$labeledType": "vegetables",
          "items": Object {
            "properties": Object {
              "veggieLike": Object {
                "$$labeledType": "veggieLike",
                "description": "Do I like this vegetable?",
                "type": "boolean",
              },
              "veggieName": Object {
                "$$labeledType": "veggieName",
                "description": "The name of the vegetable.",
                "type": "string",
              },
            },
            "required": Array [
              "veggieName",
              "veggieLike",
            ],
            "type": "object",
          },
          "type": "array",
        },
      },
      "type": "object",
    },
  ],
  "119-123": Array [
    Object {
      "$$labeledType": "veggieLike",
      "description": "Do I like this vegetable?",
      "type": "boolean",
    },
  ],
  "135-200": Array [
    Object {
      "properties": Object {
        "veggieLike": Object {
          "$$labeledType": "veggieLike",
          "description": "Do I like this vegetable?",
          "type": "boolean",
        },
        "veggieName": Object {
          "$$labeledType": "veggieName",
          "description": "The name of the vegetable.",
          "type": "string",
        },
      },
      "required": Array [
        "veggieName",
        "veggieLike",
      ],
      "type": "object",
    },
  ],
  "14-43": Array [
    Object {
      "$$labeledType": "fruits",
      "items": Object {
        "type": "string",
      },
      "type": "array",
    },
  ],
  "157-167": Array [
    Object {
      "$$labeledType": "veggieName",
      "description": "The name of the vegetable.",
      "type": "string",
    },
  ],
  "16-23": Array [
    Object {
      "type": "string",
    },
  ],
  "189-194": Array [
    Object {
      "$$labeledType": "veggieLike",
      "description": "Do I like this vegetable?",
      "type": "boolean",
    },
  ],
  "25-33": Array [
    Object {
      "type": "string",
    },
  ],
  "35-41": Array [
    Object {
      "type": "string",
    },
  ],
  "61-204": Array [
    Object {
      "$$labeledType": "vegetables",
      "items": Object {
        "properties": Object {
          "veggieLike": Object {
            "$$labeledType": "veggieLike",
            "description": "Do I like this vegetable?",
            "type": "boolean",
          },
          "veggieName": Object {
            "$$labeledType": "veggieName",
            "description": "The name of the vegetable.",
            "type": "string",
          },
        },
        "required": Array [
          "veggieName",
          "veggieLike",
        ],
        "type": "object",
      },
      "type": "array",
    },
  ],
  "67-129": Array [
    Object {
      "properties": Object {
        "veggieLike": Object {
          "$$labeledType": "veggieLike",
          "description": "Do I like this vegetable?",
          "type": "boolean",
        },
        "veggieName": Object {
          "$$labeledType": "veggieName",
          "description": "The name of the vegetable.",
          "type": "string",
        },
      },
      "required": Array [
        "veggieName",
        "veggieLike",
      ],
      "type": "object",
    },
  ],
  "89-97": Array [
    Object {
      "$$labeledType": "veggieName",
      "description": "The name of the vegetable.",
      "type": "string",
    },
  ],
}
`;

exports[`generateMenuContent 2`] = `
Array [
  Object {
    "element": Object {
      "direction": "horizontal",
      "element": Array [
        Object {
          "content": "undefined",
          "type": "display",
        },
        Object {
          "content": "remove key",
          "onSelect": Object {
            "type": "removeObjectKey",
          },
          "type": "button",
        },
      ],
      "type": "row",
    },
    "label": "TYPE",
  },
  Object {
    "element": Object {
      "content": "hi property",
      "type": "display",
    },
    "label": "PARENT",
  },
]
`;
